lang/src/org/partiql/lang/eval/physical/operators/WindowRelationalOperatorFactory.kt

```diff
@@ -0,0 +1,204 @@
+        windowExpression: PartiqlPhysical.WindowExpression,
+        windowFunctionParameter: List<ValueExpression>
+    ) = RelationExpression { state ->
+
+        // the following corresponding to materialization process
+        val source = source.evaluate(state)
+        val registers = sequence {
+            while (source.nextRow()) {
+                yield(state.registers.clone())
+            }
+        }
+
+        val partitionSortSpec = windowPartitionList?.map {
+            CompiledSortKey(NaturalExprValueComparators.NULLS_FIRST_ASC, it)
+        } ?: emptyList()
+
+        val sortKeys = partitionSortSpec + (windowSortSpecList ?: emptyList())
+
+        val sortedRegisters = registers.sortedWith(getSortingComparator(sortKeys, state))
+
+        // create the partition here
+        var partition = mutableListOf<List<Array<ExprValue>>>()
+
+        // entire partition
+        if (windowPartitionList == null) {
+            partition.add(sortedRegisters.toList())
+        }
+        // need to be partitioned
+        else {
+            val iter = sortedRegisters.iterator()
+            var rowInPartition = mutableListOf<Array<ExprValue>>()
+            var previousPartition: ExprValue? = null
+            while (iter.hasNext()) {
+                val currentRow = iter.next()
+                state.load(currentRow)
+                val currentPartition = state.valueFactory.newSexp(
+                    windowPartitionList.map {
+                        it.invoke(state)
+                    }
+                )
+
+                if (previousPartition == null) {
+                    rowInPartition.add(currentRow)
+                    previousPartition = currentPartition
+                } else if (previousPartition.exprEquals(currentPartition)) {
+                    rowInPartition.add(currentRow)
+                } else {
+                    partition.add(rowInPartition.toList())
+                    rowInPartition.clear()
+                    previousPartition = currentPartition
+                    rowInPartition.add(currentRow)
+                }
+            }
+            // finish up
+            partition.add(rowInPartition.toList())
+            rowInPartition.clear()
+        }
+
+        // We would need to model this better. This PR was for demonstrate the window operator.
+        // See the next PR for window function modeling.
+        if (windowExpression.funcName.text.toLowerCase() == "lag") {
+            LagFunction(windowExpression, partition, windowFunctionParameter, state).eval()
+        } else {
+            LeadFunction(windowExpression, partition, windowFunctionParameter, state).eval()
+        }
+    }
+}
+
+// TODO : The next PR will focus on modeling the window function.
+internal class LeadFunction(val windowExpression: PartiqlPhysical.WindowExpression, val partition: MutableList<List<Array<ExprValue>>>, val arguments: List<ValueExpression>, val state: EvaluatorState) {
```

Comment:
Like a lagLeadCommon. I tried something similar, decided to not using this approach for now since it makes the code less readable. Maybe discuss more on this matter in the next PR?